In [ ]:
NAME: HARSHA VARDHAN

***Pandas for AI/ML – Conceptual & Technical Assessment***


##SECTION-A##
**OUTPUT PREDICTION**

*QUESTION :1*...Array operations and vectorization


```text

import numpy as np
import pandas as pd

arr = np.array([10, 20, 30])
s = pd.Series([10, 20, 30])

print(arr * 2)
print(s * 2)


*OUTPUT PREDICION* :##(Q1)##


```text
[20 40 60]

0    20
1    40
2    60
dtype: int64




In [ ]:
  #(Q1)  ###CODE ANALYSIS##
#arr * 2 (numpy array):Multiple each element by 2 using fast , contionous memory llocation with explicit loops.#

#s * 2 (Pandas series): Multiple each element by 2 while preserving the explicit index alignment(0, 1, 2)#

import numpy as np
import pandas as pd

arr = np.array([10, 20, 30])
s = pd.Series([10, 20, 30])

print(arr * 2)
print(s * 2)

[20 40 60]
0    20
1    40
2    60
dtype: int64


```text

*EXPLANATION*

Both NumPy arrays and Pandas Series perform element-wise vectorized multiplication when multiplied by a scalar integer

*QESTION*
1.PREDICT THE OUTPUT OF BOTH STATEMENTS



[20 40 60]

0    20
1    40
2    60
dtype: int64


```
2.IDENTIFI ONE SIMILARITY AND ONE DIFFERENCE THE TWO DATA STRUCTURES:

similarity:both numpy arrays and pandas series support vectorized operations to perform element wise arithmetic without explicit loops.

 diference:a numpy array stores only raw numerical values, while a pandas maintain an explicit labeled index for each element.






```text
*QUSTION 2*
###- BOOLEAN FILTERING



import pandas as pd

df = pd.DataFrame({
    'Student': ['A', 'B', 'C', 'D'],
    'Marks': [45, 80, 65, 90]
})

result = df[df['Marks'] >= 65]
print(result[['Student', 'Marks']])


*OUTPUT PREDICION* :##(Q2)##

```text
Student  Marks
1       B     80
2       C     65
3       D     90




In [ ]:
#(Q2)  ###CODE ANALYSIS##
#Using pd.DataFrame creates the dataset, evaluating df['Marks'] >= 65 generates a Boolean Series mask that filters rows where marks are 65 or higher, and double brackets [['Student', 'Marks']] project the result as a DataFrame displaying only those selected columns.##

import pandas as pd

df = pd.DataFrame({
    'Student': ['A', 'B', 'C', 'D'],
    'Marks': [45, 80, 65, 90]
})

result = df[df['Marks'] >= 65]
print(result[['Student', 'Marks']])


  Student  Marks
1       B     80
2       C     65
3       D     90



```text
*EXPLANATION*
Boolean indexing filters the DataFrame by keeping only rows where the condition Marks >= 65 evaluates to True, displaying the selected columns for those matching records

*Q2*:1.PREDICT THE ROWS DISPLAYED

Student  Marks
1       B     80
2       C     65
3       D     90

2.HOW MANY ROWS AND COLUMNS WILL THE RESULTING DATA FRAME COUNT
Number of Rows: 3 (Students 'B', 'C', and 'D' satisfy the condition)   
Number of Columns: 2 (Student and Marks)   

```text
*QUESTION 3*
##DESCRIPTIVE STATISTICS##

import pandas as pd

df = pd.DataFrame({
    'Marks': [10, 20, 30, 40, 50]
})

print(df['Marks'].mean())
print(df['Marks'].max())
print(df['Marks'].median())

*OUTPUT PREDICION* :##(Q3)##

```text
30.0
50
30.0

*EXPLANATION*
This code computes and prints the average (mean), the highest value (max), and the exact middle value (median) of the Marks column.

In [ ]:
#(Q3)  ###CODE ANALYSIS##
## Code Analysis: This snippet calculates and displays the mean (30.0), maximum (50), and median (30.0) of the 'Marks' column using Pandas descriptive statistics functions.

import pandas as pd

df = pd.DataFrame({
    'Marks': [10, 20, 30, 40, 50]
})

print(df['Marks'].mean())
print(df['Marks'].max())
print(df['Marks'].median())

30.0
50
30.0



```text
*EXPLANATION*
 This code extracts basic descriptive metrics—the average (mean), highest value (max), and middle value (median)—from a numerical Pandas column

 *Q3* 1.PREDICT ALL THREE OUPUTS
 30.0
50
30.0

2.EXPLAIN HOW THE MEDIAN DIFFERS FROM THE MEAN
The mean is the mathematical average of all values, whereas the median is the central value when the dataset is sorted in order, making the median much less affected by sudden extreme outlier values.


 ```text

 *QUESTION 4*
 ##GROUP-WISE ANALYSIS##

import pandas as pd

df = pd.DataFrame({
    'Department': ['IT', 'HR', 'IT', 'HR'],
    'Salary': [40000, 30000, 60000, 50000]
})

print(df.groupby('Department')['Salary'].mean())



 ```text
*OUTPUT PREDICION* :##(Q4)##Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64




In [ ]:
#(Q4)  ###CODE ANALYSIS##
#Groupby operation 'Department' basis meeda data unique groups ga split chesi, prati group ki 'Salary' column average (mean) ni calculate chestundi.


import pandas as pd

df = pd.DataFrame({
    'Department': ['IT', 'HR', 'IT', 'HR'],
    'Salary': [40000, 30000, 60000, 50000]
})

print(df.groupby('Department')['Salary'].mean())


Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64



```text
*EXPLANATION*
df.groupby('Department')['Salary'].mean() groups the data by distinct department names and computes the average salary for each group using the Split-Apply-Combine approach.


*Q4*
1.PREDICT THE OUTPUT:

Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64

2.EXPLAIN HOW GROUPBY() CALCULATE THE DEPARTMENT WISE AVERAGE
The groupby('Department')['Salary'].mean() statement uses the Split-Apply-Combine mechanism to group rows by distinct department names and compute the average salary for each group.  
 How groupby() calculates department-wise average :Split: Pandas divides the DataFrame into distinct groups based on unique values in the 'Department' column (HR and IT):   HR Group: [30000, 50000]   IT Group: [40000, 60000]   
 Apply: The aggregation function .mean() is applied to the 'Salary' values of each group independently:   HR Mean: $(30000 + 50000) / 2 = 40000.0$   IT Mean: $(40000 + 60000) / 2 = 50000.0$   
 Combine: Pandas merges the calculated average values back into a single, labeled Pandas Series where department names serve as the index


```text
*QUESTION 5*
#*PANDAS CORRELATION-POSITIVE**

import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df['Hours'].corr(df['Marks']))






 ```text
*OUTPUT PREDICION* :##(Q5)

1.0


*EXPLANATION*

The .corr() method calculates Pearson's correlation coefficient, returning 1.0 because Marks increases linearly in direct proportion to Hours

In [ ]:
#(Q5)  ###CODE ANALYSIS##
# Code Analysis: Calculates Pearson's correlation coefficient between 'Hours' and 'Marks', returning 1.0 due to a perfect positive linear relationship.

import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df['Hours'].corr(df['Marks']))

1.0


 ```text
*Q5*
1.PREDICT THE CORRELATION COEFFICIENT?
1.0

2.INTERPRET THE DIRECTION AND STRENGTH OF THE RELATIONSHIP?
Direction: Positive correlation (as Hours increases, Marks also increases).   Strength: Perfect linear relationship (value is exactly 1.0).

 ```text

*QUESTION 6*
###PANDAS CORRELATION-NEGATIVE##

import pandas as pd

df = pd.DataFrame({
    'Price': [10, 20, 30, 40],
    'Demand': [100, 80, 60, 40]
})

print(df['Price'].corr(df['Demand']))



```text
*OUTPUT PREDICION* :##(Q6)
-1.0

*EXPLANATION*

The .corr() function calculates Pearson's correlation coefficient, returning -1.0 because Demand decreases by a constant rate ($20$ units) for every $10$-unit increase in Price


In [ ]:
#(Q5)  ###CODE ANALYSIS##
# Code Analysis: Calculates Pearson's correlation coefficient between 'Price' and 'Demand', returning -1.0 due to a perfect negative linear relationship.

import pandas as pd

df = pd.DataFrame({
    'Price': [10, 20, 30, 40],
    'Demand': [100, 80, 60, 40]
})

print(df['Price'].corr(df['Demand']))

-1.0


```text
*Q6*
1.PREDICT THE CORRELATION COEFFICIENT?
-1.0

2.EXPLAIN WHAT THE SIGN OF THE COEFFICIENT INDICATES ABOUT THE RELATIONSHIP
Sign (-): Indicates an inverse (negative) relationship—as Price increases, Demand decreases.
  Strength: Perfect linear relationship (a value of -1.0 means they move in exact opposite proportions).



******SECTION -B**********

```text
***QUESTION 1***
***JSON and CSV***

import pandas as pd

df = pd.read_csv("students.csv", header=None)
print(df)

1.A CSV file contains a header row, but the following code incorrectly reads the first student's record as column names. Correct the code



```text
###Issue Identification###
The code uses header=None, which tells Pandas that there are no column names in the file. As a result, Pandas automatically generates numeric column names (0, 1, 2...) and treats the actual header row (column names) as the first row of data.

In [ ]:
###CORRECT CODE ###
###Removing header=None restores default header=0 behavior
import pandas as pd


df = pd.read_csv("students.csv")
print(df)


    ID     Name        Date  Marks
0  101  Asha, R  2026-09-01   80.0
1  102     Ravi  01/09/2026    NaN
2  103     Neha  2026-09-03    NaN


```text
**EXPLANATION*
Removing header=None enables the default header=0 behavior, allowing Pandas to correctly parse the first line of the CSV as column headers instead of data

```text
***QUESTION 2***

###DATA ANAYSIS###

import pandas as pd

df = pd.DataFrame({
    'Employee': ['A', 'B', 'C', 'D', 'E', 'F'],
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})
print(df.sort_values('Salary').head(5))




```text
1.The objective is to display the five highest salaries from a DataFrame. Correct the code.

*Issue Identification*
By default, df.sort_values('Salary') sorts the DataFrame in ascending order (from lowest to highest). As a result, .head(5) displays the five lowest salaries instead of the five highest.   

In [ ]:
###CORRECT CODE###

import pandas as pd

df = pd.DataFrame({
    'Employee': ['A', 'B', 'C', 'D', 'E', 'F'],
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})

print(df.sort_values('Salary', ascending=False).head(5))

  Employee  Salary
5        F   70000
3        D   60000
1        B   50000
4        E   45000
0        A   30000


```text
**EXPLANATION*
Adding ascending=False sorts the DataFrame in descending order, allowing .head(5) to extract the top five highest salaries.

```text
***QUESTION 3***
CLEANING DATA###

import pandas as pd

df = pd.DataFrame({'Marks': [70, None, 90]})
df['Marks'].fillna(df['Marks'].mean())
print(df)



```text
1.A student wants to replace missing marks with the column's mean but writes code that does not update the DataFrame. Correct the code.


**Issue Identification
fillna() returns a modified copy of the Series by default and does not update the original DataFrame unless the result is reassigned or inplace=True is specified

In [ ]:
##CORRECT CODE###

import pandas as pd

df = pd.DataFrame({'Marks': [70, None, 90]})


df['Marks'] = df['Marks'].fillna(df['Marks'].mean())
print(df)

   Marks
0   70.0
1   80.0
2   90.0


```text
**EXPLANATION*
Reassigning the result of df['Marks'].fillna(...) back to df['Marks'] permanently updates the DataFrame with the imputed mean value

```text
***QUESTION 4***
###CORRELATION ###

import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})
print(df.corr('Hours', 'Marks'))


```text
1.The objective is to calculate the correlation between Hours and Marks. Correct the code.  

***Issue Identification
DataFrame.corr() accepts correlation parameters like method='pearson' instead of individual column names as positional arguments. To compute the correlation between two specific columns, you must call .corr() on one Series and pass the second Series as the argument.

In [ ]:
#CORRECT CODE###

import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df['Hours'].corr(df['Marks']))

1.0


```text
**EXPLANATION*
Calling .corr() directly on the df['Hours'] Series and passing df['Marks'] as the target argument correctly calculates their pairwise correlation.

****SECTION -C****

***EDGE CASES***


```text
***question 1***
Two NumPy arrays have shapes (3, 1) and (1, 3). Predict the shape of their element-wise sum and explain how broadcasting works. Why might the result differ from adding two one-dimensional arrays?


import numpy as np

a = np.array([[10], [20], [30]])
b = np.array([[1, 2, 3]])

result = a + b

print(result)
print(result.shape)

```text
Broadcasting stretches the (3, 1) column array and (1, 3) row array into a (3, 3) matrix to perform element-wise addition.    
1.Predict the output of the given code.
[[11 12 13]
 [21 22 23]
 [31 32 33]]
(3, 3)

*explanation*

Broadcasting stretches the (3, 1) column array and (1, 3) row array into a (3, 3) matrix to perform element-wise addition.   

```text

2.Identify the shapes of arrays a and b before addition.

Shape of array a: (3, 1)  
 Shape of array b: (1, 3)

```text

 3.Explain how NumPy broadcasting enables the addition of these arrays.

 NumPy automatically stretches dimensions with a size of 1 to match the dimension of the other array. Array a (3, 1) is stretched across columns to (3, 3), and array b (1, 3) is stretched down rows to (3, 3), enabling element-wise addition into a (3, 3) matrix.  
 * Difference from 1D Arrays: Adding two 1D arrays of shape (3,) performs direct position-by-position addition resulting in a 1D array of shape (3,), whereas adding 2D column (3, 1) and row (1, 3) arrays triggers outer expansion to form a (3, 3) 2D matrix.   

```text
4. What would happen if b were replaced with np.array([1, 2])? Explain the reason.  


Result: It raises a ValueError: operands could not be broadcast together with shapes (3,1) (2,).
   Reason: Array a has 3 rows, but b has 2 elements. Since neither dimension matches nor equals 1 along compatible axes, NumPy cannot align them for broadcasting

In [ ]:
import numpy as np

# Define arrays
a = np.array([[10], [20], [30]])
b = np.array([[1, 2, 3]])


result = a + b


print(result)
print(result.shape)



[[11 12 13]
 [21 22 23]
 [31 32 33]]
(3, 3)


```text
#Explanation:
 Broadcasting stretches the (3, 1) column vector and (1, 3) row vector into a (3, 3) matrix to perform element-wise addition.

```text

***QUESTION 2***

 A CSV file contains commas inside quoted text fields, missing values, and dates in different formats. Explain how incorrect parsing can affect the resulting DataFrame and identify suitable read options or preprocessing steps.

 Given dataset: students.csv

 ID,Name,Date,Marks
101,"Asha, R",2026-09-01,80
102,Ravi,01/09/2026,
103,Neha,2026-09-03,NA

```text
import pandas as pd

df = pd.read_csv("students.csv")
print(df)
print(df.dtypes)

```text
1.Predict how the quoted comma in "Asha, R" will be handled when the CSV file is read.

Pandas' default read_csv() parser automatically recognizes standard CSV quote rules. The double quotes around "Asha, R" instruct the parser to treat the comma inside as a literal text character rather than a column delimiter. As a result, "Asha, R" is preserved as a single string entry in the Name column without splitting it into separate columns.



```text

2.
Identify the potential issues with the Date and Marks columns.

Date Column: Contains mixed date formatting (YYYY-MM-DD like 2026-09-01 and DD/MM/YYYY like 01/09/2026). By default, Pandas imports this column as a plain text string (object dtype) instead of a proper datetime object, preventing chronological sorting, filtering, or date arithmetic.  

 Marks Column: Contains mixed data types—numerical values (80), an empty entry, and a string ("NA"). Unless "NA" is explicitly parsed as a missing value, Pandas defaults the entire column to a text string (object dtype), which prevents calculating statistics like mean, max, or median.   


```text

3.Rewrite the code to handle missing marks and convert the date column into a suitable date-time format.


In [ ]:
import pandas as pd

df = pd.read_csv("students.csv", na_values=['NA', ''])

df['Marks'] = pd.to_numeric(df['Marks'])

df['Date'] = pd.to_datetime(df['Date'], format='mixed')

print("Cleaned DataFrame:")
print(df)
print("\nData Types:")
print(df.dtypes)



Cleaned DataFrame:
    ID     Name       Date  Marks
0  101  Asha, R 2026-09-01   80.0
1  102     Ravi 2026-01-09    NaN
2  103     Neha 2026-09-03    NaN

Data Types:
ID                int64
Name                str
Date     datetime64[us]
Marks           float64
dtype: object


```text
#Explanation:
 Proper na_values parameter and pd.to_datetime with format='mixed' convert string columns into correct numeric and datetime data types.

```text
4.Explain why the value "NA" should be investigated before deciding whether it represents a missing value.  

In dataset processing, "NA" can stand for "Not Available" (a true missing value), but in academic or real-world contexts, it could also represent "Not Attempted" or "Not Applicable". If "NA" means "Not Attempted" (i.e., a score of 0), dropping or imputing it with the average mark would artificially inflate student averages and skew overall performance analysis.

***QUESTION 3***


```text
Q3.A salary column contains the values 25000, 30000, 35000, and 10000000. Explain how the extreme value affects the mean compared with the median. Identify which statistic may better represent a typical employee's salary and justify your choice

import pandas as pd

df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"],
    "Salary": [25000, 30000, 35000, 10000000]
})

print("Mean:", df["Salary"].mean())
print("Median:", df["Salary"].median())

```text

1.Predict the mean and median salary?

Mean: 2522500.0 (or $2,522,500$)   $$\text{Mean} = \frac{25000 + 30000 + 35000 + 10000000}{4} = \frac{10090000}{4} = 2522500.0$$

Median: 32500.0 (or $32,500$)   $$\text{Median} = \frac{30000 + 35000}{2} = 32500.0$$

```text
2.Identify the effect of the extreme salary on the mean.
The extreme outlier (10,000,000) heavily skews the mean upwards to over 2.5 million, making it significantly higher than 75% of the employees' actual salaries in the company.

```text
3.hich statistic may better represent a typical employee's salary in this dataset? Justify your answer.
Best Statistic: Median (32,500)   Justification: The median is robust to extreme outliers. It reflects the central tendency of the dataset without being pulled by extreme values, accurately capturing what a typical employee earns ($25k–$35k).


```text
4.Write code to identify the employee record with the highest salary and explain why the record should be investigated before removing it.



In [ ]:
import pandas as pd

df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"],
    "Salary": [25000, 30000, 35000, 10000000]
})


highest_salary_record = df.loc[df['Salary'].idxmax()]

print("Highest Salary Record:")
print(highest_salary_record)

# Explanation: The highest salary record should be investigated before deletion because it could be a legitimate executive salary or a data entry error (e.g., extra zeros added by mistake).

Highest Salary Record:
Employee           D
Salary      10000000
Name: 3, dtype: object


```text
*EXPLANATION*
Using .idxmax() locates the extreme salary record (Employee D), which requires investigation to distinguish between a legitimate executive pay scale and a data-entry error before taking deletion actions.

***SECTION -D***

***SCENARIO -BASED QUESTIONS***

```text
*QUESTION 1**
 A student assessment dataset contains missing marks, duplicate student records, inconsistent department names such as IT, it, and I.T., and marks stored as text. Develop a cleaning sequence to standardize the data, handle missing values, remove duplicates, and convert marks into a suitable numeric type. State one risk of deleting missing values without investigating why they are missing.

   Dataset: student_assessment_dirty.csv

```text

#### **Given Dataset:**student_assesment_dirty.csv
| Student_ID | Student_Name | Department | Assessment | Marks |
| :--- | :--- | :--- | :--- | :--- |
| S101 | Asha | IT | Python | 85 |
| S102 | Ravi | it | Python | 78 |
| S103 | Neha | I.T. | Python | 92 |
| S104 | Karan | HR | Python | |
| S105 | Meera | HR | Python | Absent |
| S106 | Arjun | Finance | Python | 0 |
| S107 | Priya | finance | Python | 67 |
| S108 | Rahul | IT | Python | 75 |
| S102 | Ravi | it | Python | 78 |
| S109 | Sneha | I.T. | Python | NA |
| S110 | Varun | HR | Python | 88 marks |
| S111 | Diya | Finance | Python | 91 |
| S112 | Aman | IT | Python | |
| S113 | Nisha | HR | Python | 64 |
| S114 | Kabir | Finance | Python | 105 |






1 Load the dataset and inspect it using head(), info(), and isnull().sum().



In [ ]:
import numpy as np
import pandas as pd

# Creating the dirty dataset as a DataFrame
data = {
    'Student_ID': ['S101', 'S102', 'S103', 'S104', 'S105', 'S106', 'S107', 'S108', 'S102', 'S109', 'S110', 'S111', 'S112', 'S113', 'S114'],
    'Student_Name': ['Asha', 'Ravi', 'Neha', 'Karan', 'Meera', 'Arjun', 'Priya', 'Rahul', 'Ravi', 'Sneha', 'Varun', 'Diya', 'Aman', 'Nisha', 'Kabir'],
    'Department': ['IT', 'it', 'I.T.', 'HR', 'HR', 'Finance', 'finance', 'IT', 'it', 'I.T.', 'HR', 'Finance', 'IT', 'HR', 'Finance'],
    'Assessment': ['Python'] * 15,
    'Marks': ['85', '78', '92', '', 'Absent', '0', '67', '75', '78', 'NA', '88 marks', '91', '', '64', '105']
}

df = pd.DataFrame(data)

# Inspection
print("--- First 5 Rows ---")
print(df.head())
print("\n--- Dataset Info ---")
print(df.info())
print("\n--- Missing Values Count ---")
print(df.isnull().sum())

--- First 5 Rows ---
  Student_ID Student_Name Department Assessment   Marks
0       S101         Asha         IT     Python      85
1       S102         Ravi         it     Python      78
2       S103         Neha       I.T.     Python      92
3       S104        Karan         HR     Python        
4       S105        Meera         HR     Python  Absent

--- Dataset Info ---
<class 'pandas.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Student_ID    15 non-null     str  
 1   Student_Name  15 non-null     str  
 2   Department    15 non-null     str  
 3   Assessment    15 non-null     str  
 4   Marks         15 non-null     str  
dtypes: str(5)
memory usage: 732.0 bytes
None

--- Missing Values Count ---
Student_ID      0
Student_Name    0
Department      0
Assessment      0
Marks           0
dtype: int64


```text

**EXPLANATION***

Used head(), info(), and isnull().sum() to inspect the structural layout, data types, and initial missing value counts of the dataset.


2.Standardize department names (IT, it, I.T. $\rightarrow$ IT; finance, Finance $\rightarrow$ Finance).

In [ ]:
# Mapping inconsistent department names to clean standard names
dept_map = {
    'it': 'IT',
    'I.T.': 'IT',
    'finance': 'Finance'
}

df['Department'] = df['Department'].replace(dept_map)
print("Standardized Departments:", df['Department'].unique())

Standardized Departments: <StringArray>
['IT', 'HR', 'Finance']
Length: 3, dtype: str


```text

**EXPLANATION***
Replaced variations like it and I.T. with IT, and finance with Finance using replace() to ensure consistent categorical naming.

3.Identify and remove exact duplicate records.

In [ ]:
# Identify exact duplicates and remove them
duplicates_count = df.duplicated().sum()
df = df.drop_duplicates().reset_index(drop=True)

print(f"Number of exact duplicate records removed: {duplicates_count}")

Number of exact duplicate records removed: 1


```text

**EXPLANATION***

Identified duplicate records (S102 Ravi) using duplicated() and purged them using drop_duplicates().

4.Convert the Marks column into numeric format, correctly handling entries like "88 marks", "Absent", "NA", and empty cells.

In [ ]:
# Extract numbers from string values (e.g., '88 marks' -> 88)
df['Clean_Marks'] = df['Marks'].astype(str).str.extract(r'(\d+)')[0]

# Convert extracted values to numeric (float), non-numerics become NaN
df['Marks_Numeric'] = pd.to_numeric(df['Clean_Marks'], errors='coerce')

print(df[['Student_ID', 'Student_Name', 'Marks', 'Marks_Numeric']])

   Student_ID Student_Name     Marks  Marks_Numeric
0        S101         Asha        85           85.0
1        S102         Ravi        78           78.0
2        S103         Neha        92           92.0
3        S104        Karan                      NaN
4        S105        Meera    Absent            NaN
5        S106        Arjun         0            0.0
6        S107        Priya        67           67.0
7        S108        Rahul        75           75.0
8        S109        Sneha        NA            NaN
9        S110        Varun  88 marks           88.0
10       S111         Diya        91           91.0
11       S112         Aman                      NaN
12       S113        Nisha        64           64.0
13       S114        Kabir       105          105.0


```text

**EXPLANATION***
Extracted numeric digits from strings like "88 marks" using regular expression str.extract(r'(\d+)') and converted them to numbers using pd.to_numeric().

5 Distinguish absent students from missing entries and state the risk of deleting missing values without investigation.

In [ ]:
# Distinguish between Absent students and missing/unentered marks
df['Status'] = np.where(df['Marks'] == 'Absent', 'Absent',
               np.where(df['Marks_Numeric'].isna(), 'Missing', 'Present'))

print(df[['Student_ID', 'Student_Name', 'Marks_Numeric', 'Status']])

   Student_ID Student_Name  Marks_Numeric   Status
0        S101         Asha           85.0  Present
1        S102         Ravi           78.0  Present
2        S103         Neha           92.0  Present
3        S104        Karan            NaN  Missing
4        S105        Meera            NaN   Absent
5        S106        Arjun            0.0  Present
6        S107        Priya           67.0  Present
7        S108        Rahul           75.0  Present
8        S109        Sneha            NaN  Missing
9        S110        Varun           88.0  Present
10       S111         Diya           91.0  Present
11       S112         Aman            NaN  Missing
12       S113        Nisha           64.0  Present
13       S114        Kabir          105.0  Present



```text
Risk of Deleting Missing Values Without Investigation:
Deleting missing values blindly introduces sample bias by removing legitimate records (such as students who were absent vs. those whose records failed to record due to a system glitch), ultimately skewing descriptive metrics like the overall mean.

6 Identify marks outside the valid range of 0–100 and decide on their handling.

In [ ]:
# Identify marks outside 0 to 100
invalid_mask = (df['Marks_Numeric'] < 0) | (df['Marks_Numeric'] > 100)
invalid_count = invalid_mask.sum()

print("Invalid Mark Records Found:")
print(df[invalid_mask][['Student_ID', 'Student_Name', 'Marks_Numeric']])

# Flagging out of range values as NaN for re-verification
df.loc[invalid_mask, 'Marks_Numeric'] = np.nan

Invalid Mark Records Found:
   Student_ID Student_Name  Marks_Numeric
13       S114        Kabir          105.0


```text

**EXPLANATION***

Detected out-of-range marks (e.g., 105 for Kabir) using boolean indexing and set them to NaN for further investigation.


7.Display the final cleaned DataFrame along with a summary report of records removed, missing marks remaining, and invalid marks identified.

In [ ]:
# Final dataset cleanup
cleaned_df = df[['Student_ID', 'Student_Name', 'Department', 'Assessment', 'Marks_Numeric', 'Status']].rename(
    columns={'Marks_Numeric': 'Marks'}
)

print("================ SUMMARY REPORT ================")
print(f"1. Duplicate Records Removed    : {duplicates_count}")
print(f"2. Remaining Missing Marks (NaN) : {cleaned_df['Marks'].isna().sum()}")
print(f"3. Invalid Marks Identified     : {invalid_count}")
print("================================================\n")

print("Final Cleaned DataFrame:")
print(cleaned_df)

================ SUMMARY REPORT ================
1. Duplicate Records Removed    : 1
2. Remaining Missing Marks (NaN) : 5
3. Invalid Marks Identified     : 1

Final Cleaned DataFrame:
   Student_ID Student_Name Department Assessment  Marks   Status
0        S101         Asha         IT     Python   85.0  Present
1        S102         Ravi         IT     Python   78.0  Present
2        S103         Neha         IT     Python   92.0  Present
3        S104        Karan         HR     Python    NaN  Missing
4        S105        Meera         HR     Python    NaN   Absent
5        S106        Arjun    Finance     Python    0.0  Present
6        S107        Priya    Finance     Python   67.0  Present
7        S108        Rahul         IT     Python   75.0  Present
8        S109        Sneha         IT     Python    NaN  Missing
9        S110        Varun         HR     Python   88.0  Present
10       S111         Diya    Finance     Python   91.0  Present
11       S112         Aman         I

```text

**EXPLANATION***

Explanation: Outputted the final structured DataFrame alongside a concise summary report detailing records removed, missing values, and invalid entries.